Schema based ClojureScript keyword <-> TypeScript property mapping

flow-jacques opened the pull request.

flow-jacques committed:
WIP

flow-jacques committed:
All `to-clj` unit tests are passing

flow-jacques committed:
Test to map to a single JS object are passing

flow-jacques committed:
Some test for `to-js` are passing

flow-jacques committed:
All tests are passing

flow-jacques committed:
Added a more idiomatic `toClj` convenience func for TS

flow-jacques committed:
WIP

flow-jacques committed:
Can now pass a schema directly to to-clj/js,without registry lookup

flow-danny commented on src/malli_ts/data_mapping.cljc:
```diff
@@ -0,0 +1,210 @@
+              (when (seq clj-keys->js-props)
+                (swap! *result concat clj-keys->js-props))))
+          schema))
+       (let [clj-keys->js-props (into {} @*result)
+             js-props->clj-keys (set/map-invert clj-keys->js-props)]
```
Do you have an example what `clj-keys->js-props` looks like?
If it's simply inverted won't we get naming collisions as javascript doesn't have namespaced keywords?

For example if you have
```clojure
[:map [:person/name string?]]
```
and
```clojure
[:map [:company/name string?]]
```
mapping them simply to `"name"` in JS will be lossy.

If `clj-keys->js-props` is a map from `schema -> keyword -> js property`
then `js-props->clj-keys` can be a map from `schema -> js property name -> keyword` without different map types colliding with eachother.

`map-proxy` then needs also needs to know the schema for the level it's proxying so that it can correctly map the keywords, but also do value conversions according to schema.

flow-jacques replied:
Example mapping:
```
{:clj-keys->js-props
 {:order-type "orderType",
  :amount "amount",
  :order-item "orderItem",
  :order-items "orderItems",
  :name "name",
  :test-dummy "TESTDummyXYZ",
  :type "type",
  :user-id "userId",
  :currency "currency",
  :order-id "orderId",
  :order-item-id "orderItemId",
  :model-type "modelType",
  :valid-for-timespan "validForTimespan",
  :how-is-related "howIsRelated",
  :total-amount "totalAmount",
  :price "price",
  :credit "credit",
  :user "user",
  :related-items "relatedItems"},
 :js-props->clj-keys
 {"orderItems" :order-items,
  "howIsRelated" :how-is-related,
  "totalAmount" :total-amount,
  "orderItem" :order-item,
  "TESTDummyXYZ" :test-dummy,
  "relatedItems" :related-items,
  "orderItemId" :order-item-id,
  "modelType" :model-type,
  "currency" :currency,
  "credit" :credit,
  "user" :user,
  "name" :name,
  "userId" :user-id,
  "type" :type,
  "amount" :amount,
  "orderId" :order-id,
  "price" :price,
  "validForTimespan" :valid-for-timespan, 
  "orderType" :order-type}}
```

flow-danny commented on src/malli_ts/data_mapping.cljc:
```diff
@@ -0,0 +1,210 @@
+       data
+       (js/Proxy. data
+                  #js
+                  {:get            (partial map-proxy-get js-props->clj-keys)
```
See above:
if `js-props->clj-keys` is flat, this means every object has to use unique named keys, correct?

flow-jacques replied:
@flow-danny yes, `js-props->clj-keys` is flat per root schema. One of the reasons is because `bean`'s `:prop->key` and `:key->prop` only gives you a key/property name, but it doesn't tell you where in the hierarchy you are, for example by giving you a path (in the case of nested maps). So you're forced to look the mapping up in a flat index for all keys/props across the root schema/map's keys/properties.

flow-jacques committed:
WIP

flow-jacques committed:
`clj<->js-key-mapping2` that maps from props to keys

flow-jacques committed:
New version of `to-clj` that passes schema to transform is working

flow-jacques committed:
js->clj-mapping now works with deep nesting + refs

flow-jacques committed:
WIP: `to-clj` is working now with new nested data mapping

flow-jacques committed:
`to-js` mapping basic scenarios has been fixed

flow-jacques committed:
Unit tests are working again

flow-jacques committed:
data-mapping ns split into to-js and to-clj

flow-danny committed:
Update to cljs-bean fork master

flow-jacques committed:
`shadow-cljs compile test` fixed (core-test still fails though)

flow-danny committed:
Support `clj -M:shadow-cljs watch test`

fix JS data_mapping failing to compile to Clojure (also no use for it)

flow-danny committed:
Fix import path and sorting in ClojureScript, units test pass now

flow-danny committed:
GitHub: run tests also in shadow-cljs

flow-danny committed:
Add simple-benchmark test comparing mapped JS to regular Clojure map

flow-danny committed:
Memoize clj<->js-mapping

Before:
```
[], (a-regular-clj-object), 10000 runs, 458 msecs
[], (test-a-js-obj-to-clj), 10000 runs, 6915 msecs
```

After:
```
[], (a-regular-clj-object), 10000 runs, 455 msecs
[], (test-a-js-obj-to-clj), 10000 runs, 795 msecs
```

flow-danny committed:
Faster clj<->js-mapping

Before:
```
[], (a-regular-clj-object), 10000 runs, 455 msecs
[], (test-a-js-obj-to-clj), 10000 runs, 795 msecs
```

After:
```
[], (a-regular-clj-object), 10000 runs, 445 msecs
[], (test-a-js-obj-to-clj), 10000 runs, 752 msecs
```

flow-danny committed:
Faster to-clj

Before:
```
[], (a-regular-clj-object), 10000 runs, 454 msecs
[], (test-a-js-obj-to-clj), 10000 runs, 778 msecs
```

After:
```
[], (a-regular-clj-object), 10000 runs, 450 msecs
[], (test-a-js-obj-to-clj), 10000 runs, 584 msecs
```

flow-danny committed:
Shave another comparison from to-clj

About 20ms runtime reduction:
```
[], (a-regular-clj-object), 10000 runs, 454 msecs
[], (test-a-js-obj-to-clj), 10000 runs, 562 msecs
```

flow-danny committed:
Shrink to-clj and add support JsProxy in nested bean transform

flow-danny committed:
Use direct field access into malli-ts.data-mapping/Mapping record

Before:
```
[], (a-regular-clj-object), 10000 runs, 455 msecs
[], (test-a-js-obj-to-clj), 10000 runs, 567 msecs
```

After:
```
[], (a-regular-clj-object), 10000 runs, 454 msecs
[], (test-a-js-obj-to-clj), 10000 runs, 554 msecs
```

flow-danny committed:
Getting very close!

flow-danny committed:
Failed optimization experiment: generate BeanContexts in schema walker

Before, with stable 99.1 MB memory usage:
```
========= Running Tests =======================
[], (a-regular-clj-object), 200000 runs, 3896 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4142 msecs

[], (a-regular-clj-object), 200000 runs, 3825 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4115 msecs

[], (a-regular-clj-object), 200000 runs, 3847 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4104 msecs

[], (a-regular-clj-object), 200000 runs, 3825 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4103 msecs

[], (a-regular-clj-object), 200000 runs, 3823 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4108 msecs

[], (a-regular-clj-object), 200000 runs, 3840 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4181 msecs

[], (a-regular-clj-object), 200000 runs, 3983 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4146 msecs

[], (a-regular-clj-object), 200000 runs, 3835 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4135 msecs

[], (a-regular-clj-object), 200000 runs, 3840 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4135 msecs
```

After with stable 110.4 MB memory usage:
```
========= Running Tests =======================
[], (a-regular-clj-object), 200000 runs, 3848 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4151 msecs

[], (a-regular-clj-object), 200000 runs, 3821 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4120 msecs

[], (a-regular-clj-object), 200000 runs, 3836 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4178 msecs

[], (a-regular-clj-object), 200000 runs, 3999 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4207 msecs

[], (a-regular-clj-object), 200000 runs, 3897 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4142 msecs

[], (a-regular-clj-object), 200000 runs, 3826 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4132 msecs

[], (a-regular-clj-object), 200000 runs, 3846 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4136 msecs

[], (a-regular-clj-object), 200000 runs, 3828 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4136 msecs

[], (a-regular-clj-object), 200000 runs, 3839 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4153 msecs
```

flow-danny committed:
Partially revert "Failed optimization experiment"

This reverts commit 16a244fe8b85ddc795314c6cba2a1a5768be92cd.

Before this revert:
```
========= Running Tests =======================
[], (a-regular-clj-object), 200000 runs, 3860 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4142 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4726 msecs

[], (a-regular-clj-object), 200000 runs, 3767 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4061 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4736 msecs

[], (a-regular-clj-object), 200000 runs, 3874 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4260 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4744 msecs

[], (a-regular-clj-object), 200000 runs, 3793 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4069 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4687 msecs

[], (a-regular-clj-object), 200000 runs, 3784 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4073 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4697 msecs

[], (a-regular-clj-object), 200000 runs, 3786 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4074 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4683 msecs
```

After this revert:
```
========= Running Tests =======================
[], (a-regular-clj-object), 200000 runs, 3818 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4008 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4537 msecs

[], (a-regular-clj-object), 200000 runs, 3752 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4020 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4543 msecs

[], (a-regular-clj-object), 200000 runs, 3772 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4237 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4711 msecs

[], (a-regular-clj-object), 200000 runs, 3772 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4054 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4615 msecs

[], (a-regular-clj-object), 200000 runs, 3756 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4011 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4549 msecs

[], (a-regular-clj-object), 200000 runs, 3778 msecs
[], (test-a-js-obj-to-clj), 200000 runs, 4011 msecs
[], (test-a-clj-map-to-js), 200000 runs, 4571 msecs
```

flow-danny committed:
Do only 7 benchmark runs (for GitHub actions)

flow-danny committed:
to-js optimization experiments

flow-danny committed:
to-js: remove slowest experiments

flow-danny committed:
Some cleanup

flow-danny committed:
unchecked get/set/inc optimizations

flow-danny committed:
Make JS mapping unit-test work with advanced compilation

flow-danny committed:
Don't crash on accessing unmapped (undefined) property
